Source/Core/InputCommon/GCAdapter.cpp

```diff
@@ -92,20 +95,23 @@ enum class ControllerType : u8
   Wireless = 2,
 };
 
-struct PortState
-{
-  ControllerType controller_type = ControllerType::None;
-};
-
-static std::array<PortState, SerialInterface::MAX_SI_CHANNELS> s_port_states;
 static std::array<u8, SerialInterface::MAX_SI_CHANNELS> s_controller_rumble;
 
 constexpr size_t CONTROLER_INPUT_PAYLOAD_EXPECTED_SIZE = 37;
 constexpr size_t CONTROLER_OUTPUT_INIT_PAYLOAD_SIZE = 1;
 constexpr size_t CONTROLER_OUTPUT_RUMBLE_PAYLOAD_SIZE = 5;
 
+struct PortState
+{
+  GCPadStatus origin = {};
+  GCPadStatus status = {};
+
+  ControllerType controller_type = ControllerType::None;
+  bool is_new_connection = false;
+};
```

Comment:
I'm not completely sure I understand what the `origin` field is used for. It roughly seems like it's only used once, and in that situation `status` wouldn't have been used. What problem is it trying to solve?

Reply:
It solves the future problem of ControllerInterface reading inputs simultaneously with `CSIDevice_GCAdapter`. `CSIDevice_GCAdapter` needs the "origin" after the "controller_type" is reset so it needs to be saved.
In my ControllerInterface GCAdapter PR I'm adding a `PeekInput` function to read `status` without changing the `is_new_connection` flag.
The whole `GET_ORIGIN` thing is a mess, but it's all tied into NetPlay and I'm trying to avoid touching all that fragile code.

Reply:
That sounds reasonable (though I'm still not entirely sure *why* it would need to care about a specific origin value - if the stick isn't centered on startup, bad things are always going to happen, and the same would make sense if the user immediately messed with it after startup I think...)

Reply:
Some games seem to actually use the "origin" value and can function properly with a non-centered stick on attachment (Mario Kart Double Dash seems to be one).
And allegedly the controller "origin" can be read from the GCAdapter but we aren't doing that yet: https://bugs.dolphin-emu.org/issues/12956
It's a bit strange now to store the origin separately when it's currently just a random pad state but It should make more sense if the issue's suggestion is implemented.